Second try; PR for review

Manishearth opened the pull request:
cc @Munksgaard

Munksgaard committed:
Started rewriting lib.rs

I've somewhat started over. I feel like this is a bit cleaner, and it
readily handles function calls and method calls alike.

Also handles if branches

Still very much WIP

Munksgaard committed:
Rewrote test to reflect changes

Manishearth commented on src/lib.rs:
```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
-                        protected += 1;
-                        return;
+                    if ty::has_attr(self.cx.tcx, did, "drop_protect") {
+                        self.cx.tcx.sess.span_note(pat.span, "Adding drop protected type to map");
```
Note: You want to make sure you don't do this twice, else the same nodeid for initialization will be used for multiple protected variables.

Eg 

`let (x,y) = foo();`, where `x` and `y` are both protected.

This might get more complicated when we have structs containing multiple protected types. Hopefully we don't have these.

We can ignore this for now but it is something we'll need to eventually address somehow for correctness.

Munksgaard replied:
I'm thinking we should actually use walk_pat instead. In your example, we'd need to track both `x` and `y`, but the current implementation tracks `(x,y)` as a whole.

Munksgaard replied:
I also think this is also necessary for `match` to work properly.

Manishearth replied:
Note that we can only insert one of them into the hashmap (it's keyed by nodeids, and the whole pat is a single node with no subnodes IIRC since pats themselves aren't nodes).

Manishearth commented on src/lib.rs:
```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
+                visit::walk_expr(self, e);
+            }
+            ExprAddrOf(_, _) => {
+                // Ignore children
```
Not so simple, sadly `&({x = y; z})` works.

We want to determine if this is a simple `ExprAddrOf` (i.e., one that just contains an `ExprPath`), else recurse.

Munksgaard replied:
Ah, good catch.

Munksgaard replied:
Maybe we can just track borrows as well for now?

Munksgaard replied:
Meaning, delete the whole ExprAddrOf pattern.

Manishearth replied:
Perhaps. But it would be easier to simply do what I listed above:
- If the `expr` inside the `ExprAddrOf` is an `ExprPath`, then it's something simple like `&x`, which we can ignore.
- If it's more complicated, recurse. Worst case scenario we count a borrow as a move, and emit an additional error. Not a big deal IMO.

Munksgaard replied:
Alright, I'll try it out.

Munksgaard replied:
Check out 512d68b

Manishearth commented on src/lib.rs:
```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
-                    }
+                // TODO: Add all protected vars in lhs
+                if let ExprPath(_, _) = e1.node {
+                    self.map.insert(expr_to_localid(self.cx.tcx, e1).unwrap(), e.span);
```
I'm not sure why this is necessary.

Is this for deferred lets? (`let x; ...; x = 5;`) In that case disallowing deferred lets with protected types sounds easier (and deferred lets are rarely used). 

Perhaps I'm misunderstanding what this does

Munksgaard replied:
My test-case was something like `let mut x = Foo; x = x`, variants of which we use quite a bit in our session code.

Manishearth commented on src/lib.rs:
```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
+                    let mut v2 = self.clone();
+                    v1.visit_block(&b);
+                    v2.visit_expr(&e2);
+                    if v1.map != v2.map {
```
So simple; I should have thought of that! :smile: 

Might lead to lots of churn but meh, compile time perf is okay.

Also, be careful on how equality on hashmaps is calculated, it might not be what we expect.

Munksgaard replied:
I'll keep an eye out for that.

Manishearth commented on src/lib.rs:
```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
+                        self.cx.tcx.sess.span_err(e.span, "If branch is not linear");
+                    }
+                    // Is this necessary?
+                    // self.map = v1.map;
```
Either way works. We don't need to worry too much about making the lint continue smoothly after it's found an error.

Manishearth commented:
This doesn't seem to track the blocks, but come to think of it do we really need to? Tracking blocks lets us know when something is implicitly dropped, but checking the emptiness of the map works too.

Manishearth commented:
Also, I wonder how well it handles constructs like:

```
struct Composite {
 x : Foo // protected
 // ...
}
```

with partial moves and whatnot.

We could of course mandate that structs containing protected variables be themselves marked as protected, which makes life easier.

Manishearth commented:
Overall, looks good! I may have another poke at this tomorrow to check how durable it is, but feel free to merge!

Munksgaard commented:
> This doesn't seem to track the blocks, but come to think of it do we really need to? Tracking blocks lets us know when something is implicitly dropped, but checking the emptiness of the map works too.

That's what I figured. I don't think we need block ids at the moment.

Munksgaard committed:
Make is_protected a method on the visitor

Munksgaard committed:
Prettify error output a bit

Munksgaard committed:
Use walk_block instead of visit_block

Munksgaard committed:
Rewrite of walk_pat_and_add

This is what we do:
 - Walk the pattern until we find a PatIdent, that is, the user hasn't
 destructured the value any more. Then for each such ident, we determine
 if it is to be protected, by walking it's type. That way, stuff like
 `x: Option<Foo>` is protected, but in something like `match Some(Foo) {
 Some(x) => ...}` only the `x` is protected, and not `Some(x)`.

Munksgaard committed:
Better removal of ExprPath

If something is an ExprPath, we simply remove that path's defid if it is
in our map.

Munksgaard committed:
We should use visit_expr inside visit_expr

Also, remove some unnecessary derefs

Munksgaard committed:
Add support for `match` statements

TODO: Handle guards

Munksgaard commented:
One thing we don't handle is functions like this:

``` rust
fn dropit<T>(x: T) { }
```

Since we just check all functions, but not instantiations of them, we can call `dropit` on a drop protected var. Any suggestions?

Manishearth commented:
We can check at the call site.

Munksgaard commented:
But check what? We don't annotate functions that can receive protected vars any more. We could check that a similar type is returned, but how do we know if it's the same var as before? Or that it hasn't been wrapped or unwrapped somehow?

Manishearth commented:
But we do; we still have `allow_drop` on the functions, right? If one of the arguments transitively resolves to a protected type (unless behind `&`, we probably want a utility function for that), and the function isn't annotated with `#[allow_drop(Foo)]`, then it's a case of a generic function being fed a protected type, and we can error.

Munksgaard committed:
Check that ExprAssign doesn't overwrite protected

Munksgaard committed:
Better handling of & expressions

Munksgaard commented:
We also need a better way to test all this... Something like rustc's compiletest would be perfect

Munksgaard committed:
Preliminary support for for loops

We still need to handle breaks and such in a proper manner

Munksgaard committed:
add for loop test

Munksgaard committed:
Add some simple examples

Munksgaard commented:
> But we do; we still have allow_drop on the functions, right? If one of the arguments transitively resolves to a protected type (unless behind &, we probably want a utility function for that), and the function isn't annotated with #[allow_drop(Foo)], then it's a case of a generic function being fed a protected type, and we can error.

Hmm, but then we'd have problems with vectors, right? Stuff like `let mut v = Vec::new(); let c = Chan(); v.push(c);` should be allowed.

Manishearth commented:
> Hmm, but then we'd have problems with vectors, right? Stuff like `let mut v = Vec::new(); let c = Chan(); v.push(c);` should be allowed.

Oh, right, we removed the checks on `ExprFnCall` and whatnot. We could add some sort of check that ensures that a protected type is never accepted by an unannotated generic, perhaps?
